### PR TITLE
ghc 9.6.2

### PR DESCRIPTION
**Pull request by iMichka**
- [x] allureofthestars https://github.com/AllureOfTheStars/Allure/issues/125
- [x] git-annex https://github.com/aristidb/aws/issues/288, #144920
- [x] hopenpgp-tools https://salsa.debian.org/clint/hopenpgp-tools/-/issues/5, #144657
- [x] pandoc-crossref https://github.com/lierdakil/pandoc-crossref/issues/405
- [x] rome https://github.com/tmspzz/Rome/issues/262, #148467

### CODE REVIEW

**Comment by iMichka** [comment]
I opened an issue for `allureofthestars`: https://github.com/AllureOfTheStars/Allure/issues/125

File: Formula/g/ghc.rb [hunk 1]
```diff
@@ -38,73 +29,54 @@ class Ghc < Formula
   depends_on "autoconf" => :build
   depends_on "automake" => :build
   depends_on "python@3.11" => :build
-  depends_on "sphinx-doc" => :build
-  depends_on macos: :catalina
+  depends_on "xz" => :build
+  depends_on "libffi"
```
Comment (Bo98) [inline Formula/g/ghc.rb:33]: I'm not seeing where this is used

```
   /opt/homebrew/Cellar/ghc/9.6.2/lib/ghc-9.6.2/lib/aarch64-osx-ghc-9.6.2/libffi.dylib (ghc)
Dependencies with no linkage:
  libffi
```

File: Formula/g/ghc.rb [hunk 2]
```diff
@@ -147,7 +119,7 @@ def install
       -j#{ENV.make_jobs}
       --prefix=#{prefix}
       --flavour=release
-      --docs=no-sphinx-pdfs
+      --docs=no-sphinx
```
Comment (Bo98) [inline Formula/g/ghc.rb:122]: Does this mean no more manpages? Is it broken?

You may want to check the "sphinx-doc 7" patch in `ghc@9.2` (the sphinx-doc 6 one should already be in included in the release).

File: Formula/g/ghc.rb [hunk 3]
```diff
@@ -38,73 +29,54 @@ class Ghc < Formula
   depends_on "autoconf" => :build
   depends_on "automake" => :build
   depends_on "python@3.11" => :build
-  depends_on "sphinx-doc" => :build
-  depends_on macos: :catalina
+  depends_on "xz" => :build
```
Comment (Bo98) [inline Formula/g/ghc.rb:32]: Is this not an implicit dependency from the URL being `.xz`?

File: Formula/c/cabal-install.rb [hunk 4]
```diff
@@ -1,20 +1,10 @@
...
-    sha256 "61ce436f2e14e12bf07ea1c81402362f46275014cd841a76566f0766d0ea67e6"
-
-    # Use Hackage metadata revision to support GHC 9.4.
-    # TODO: Remove this resource on next release along with corresponding install logic
-    resource "cabal-install.cabal" do
-      url "https://hackage.haskell.org/package/cabal-install-3.8.1.0/revision/2.cabal"
-      sha256 "e29a58254bb8aaf950bf541e0fe9cf63f9ae99b8ae1f7f47b62b863c25dd54d0"
-    end
-  end
+  head "https://github.com/haskell/cabal.git", branch: "3.1O"
```
Comment (Bo98) [inline Formula/c/cabal-install.rb:7]:
```suggestion
  head "https://github.com/haskell/cabal.git", branch: "3.10"
```

**Review by Bo98 — COMMENTED** [review]
Some lesser important comments:

**Comment by Bo98** [comment]
As for the build failures: we do deliberately delete that file, but the `ghc-pkg recache` in the post_install is supposed to recreate it. Maybe that's not running properly?

**Comment by Bo98** [comment]
Oh right I see the issue with postinstall. It seems to run it at the end for some odd reason. Will take a look why that is.

**Comment by Bo98** [comment]
https://github.com/Homebrew/homebrew-test-bot/pull/968

File: Formula/g/ghc.rb (hunk 1 above)
Comment (iMichka) [reply Formula/g/ghc.rb]: I've removed it, it was added in #137249, but I don't know why ... let's try without it

File: Formula/g/ghc.rb (hunk 3 above)
Comment (iMichka) [reply Formula/g/ghc.rb]: In the previous ghc build attempt they hit the  `Linux: /bin/sh: 1: xz: not found ` issue, so I copied the dependency from that PR. Let met try without.

Comment (Bo98) [reply Formula/g/ghc.rb]: Ah ok, the build itself maybe uses it somewhere. Make sense, thanks for clarifying.

It used to be the case that implicit dependencies were also in the build environment but we changed that recently.

Comment (Bo98) [reply Formula/g/ghc.rb]: It's possible it might only be needed on Linux, so worth seeing what the CI says.

**Comment by iMichka** [comment]
> | Copy file: docs/index.html => _build/doc/html/index.html
  | Run Tar Create: _build/doc/html/libraries => _build/doc/archives/libraries.html.tar.xz
  Command line: /usr/bin/tar -c --xz -f _build/doc/archives/libraries.html.tar.xz _build/doc/html/libraries
  ===> Command failed with error code: 2
  /bin/sh: 1: xz: not found
  /usr/bin/tar: _build/doc/archives/libraries.html.tar.xz: Wrote only 4096 of 10240 bytes
  /usr/bin/tar: Child returned status 127
  /usr/bin/tar: Error is not recoverable: exiting now

On Linux. So `xz` is necessary.

It does get installed:
> Full install --build-from-source ghc output
  ==> Fetching dependencies for ghc: xz, m4, ca-certificates, openssl@3, berkeley-db, gdbm, expat, libxcrypt, perl, autoconf, automake, mpdecimal, ncurses, readline, zlib, sqlite, bzip2, libedit, libffi, unzip, berkeley-db@5, krb5, libtirpc, libnsl, python@3.11 and gmp

But you still need to declare it as a build dependency?

File: Formula/g/ghc.rb (hunk 2 above)
Comment (iMichka) [reply Formula/g/ghc.rb:120]: I will add it back with the patches

**Comment by iMichka** [comment]
> 2023-09-17T07:40:00.4459000Z Command line: _build/stage0/bin/ghc -Wall -Wcompat -fdiagnostics-color=never -hisuf p_hi -osuf p_o -hcsuf p_hc -static -prof -hide-all-packages -no-user-package-db '-package-env -' '-package-db _build/stage1/inplace/package.conf.d' '-this-unit-id ghc-prim-0.10.0' '-package-id rts-1.0.2' -i -i/private/tmp/ghc-20230917-29440-470b1d/ghc-9.6.2/_build/stage1/libraries/ghc-prim/build -i/private/tmp/ghc-20230917-29440-470b1d/ghc-9.6.2/_build/stage1/libraries/ghc-prim/build/autogen -i/private/tmp/ghc-20230917-29440-470b1d/ghc-9.6.2/libraries/ghc-prim -Irts/include -I_build/stage1/libraries/ghc-prim/build -I/private/tmp/ghc-20230917-29440-470b1d/ghc-9.6.2/rts/include -I/private/tmp/ghc-20230917-29440-470b1d/ghc-9.6.2/_build/stage1/rts/build/include -I/private/tmp/ghc-20230917-29440-470b1d/ghc-9.6.2/rts/@FFIIncludeDir@ -I/private/tmp/ghc-20230917-29440-470b1d/ghc-9.6.2/_build/stage1/rts/build/@FFIIncludeDir@ -I/private/tmp/ghc-20230917-29440-470b1d/ghc-9.6.2/rts/@LibdwIncludeDir@ -I/private/tmp/ghc-20230917-29440-470b1d/ghc-9.6.2/_build/stage1/rts/build/@LibdwIncludeDir@ -optP-include -optP_build/stage1/libraries/ghc-prim/build/autogen/cabal_macros.h -optc--target=arm64-apple-darwin -outputdir _build/stage1/libraries/ghc-prim/build -Wnoncanonical-monad-instances -optc-Wno-unknown-pragmas -optP-Wno-nonportable-include-path -c libraries/ghc-prim/GHC/Types.hs -o _build/stage1/libraries/ghc-prim/build/GHC/Types.p_o -O -H64m -this-unit-id ghc-prim -XHaskell2010 -no-global-package-db -package-db=/private/tmp/ghc-20230917-29440-470b1d/ghc-9.6.2/_build/stage1/inplace/package.conf.d -ghcversion-file=rts/include/ghcversion.h -ghcversion-file=rts/include/ghcversion.h -O2 -Wno-deprecated-flags -Wno-trustworthy-safe -haddock
2023-09-17T07:40:00.4462070Z ===> Command failed with error code: 1
2023-09-17T07:40:00.4462440Z clang: warning: -Wl,-headerpad_max_install_names: 'linker' input unused [-Wunused-command-line-argument]
2023-09-17T07:40:00.4463070Z clang: warning: argument unused during compilation: '-L/opt/homebrew/lib' [-Wunused-command-line-argument]
2023-09-17T07:40:00.4463580Z clang: warning: argument unused during compilation: '-L/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries' [-Wunused-command-line-argument]
2023-09-17T07:40:00.4463870Z 
2023-09-17T07:40:00.4463940Z /private/tmp/ghc47071_0/ghc_4.c:4:10: error:
2023-09-17T07:40:00.4464270Z      fatal error: 'ffi.h' file not found
2023-09-17T07:40:00.4464440Z   |
2023-09-17T07:40:00.4464560Z 4 | #include <ffi.h>
2023-09-17T07:40:00.4464680Z   |          ^
2023-09-17T07:40:00.4464810Z #include <ffi.h>
2023-09-17T07:40:00.4464930Z          ^~~~~~~
2023-09-17T07:40:00.4465230Z 1 error generated.
2023-09-17T07:40:00.4465750Z `clang' failed in phase `C Compiler'. (Exit code: 1)
2023-09-17T07:40:00.4465950Z Command failed
2023-09-17T07:40:00.4466540Z Build failed.
2023-09-17T07:40:00.4466630Z 
2023-09-17T07:40:00.4466850Z [31mREAD THIS:[0m [4mhttps://docs.brew.sh/Troubleshooting[24m

Here is why libffi was necessary

**Comment by Bo98** [comment]
> Here is why libffi was necessary

That's because we're configuring it wrong: https://gitlab.haskell.org/ghc/ghc/-/issues/22595#note_468423

`--with-ffi-libraries=#{MacOS.sdk_path_if_needed}/usr/lib --with-ffi-includes=#{MacOS.sdk_path_if_needed}/usr/include/ffi` should fix it

We generally prefer Apple's libffi as it has arm64 trampoline support which ours does not.

**Comment by Bo98** [comment]
> On Linux. So `xz` is necessary.
> 
> It does get installed:
> 
> > Full install --build-from-source ghc output
> > ==> Fetching dependencies for ghc: xz, m4, ca-certificates, openssl@3, berkeley-db, gdbm, expat, libxcrypt, perl, autoconf, automake, mpdecimal, ncurses, readline, zlib, sqlite, bzip2, libedit, libffi, unzip, berkeley-db@5, krb5, libtirpc, libnsl, python@3.11 and gmp
> 
> But you still need to declare it as a build dependency?

Yeah it's downloaded to extract the main archive but removed from the build deps when building. It's a change we made recently. So yeah all good to add it back as a Linux-only dependency.

**Comment by Bo98** [comment]
> `configure: WARNING: --with-ffi-includes will be ignored, --with-system-libffi not set`

ah oops need `--with-system-libffi` as well. Should probably set that on all platforms.

**Comment by clinty** [comment]
Does https://salsa.debian.org/clint/hopenpgp-tools/-/commit/fc4214399f06d4ddeb2ecf93ddd3d9bc9ed140bc work for you?

**Comment by iMichka** [comment]
> Does https://salsa.debian.org/clint/hopenpgp-tools/-/commit/fc4214399f06d4ddeb2ecf93ddd3d9bc9ed140bc work for you?

Yes that worked, thanks! I'll fix this in a separate pull request #144657 for homebrew.

**Comment by fishtreesugar** [comment]
FYI, [GHC 9.6.3 released](https://www.haskell.org/ghc/blog/20230925-ghc-9.6.3-released.html)

**Comment by iMichka** [comment]
> brew readall --aliases --os=all --arch=all homebrew/core
Error: Non-file alias: /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Aliases/ghc@9.6

**Comment by fxcoudert** [comment]
haskell-language-server checksum mismatch

**Comment by iMichka** [comment]
That's odd. I still had the checksum in the terminal history and I think I did nothing wrong. I redownloaded the file right now and it is still 7123eabb149bd0a392f3576b5383efb561600c51e361f27e424dbd0d71ed1de0. Maybe the checksum depends from where you download the file ... or it might change over time, as it is an autogenerated file by GitHub.

Nevertheless I have removed the CI-no-fail fast, we need a faster feedback loop now that we are close to completion.

**Comment by fxcoudert** [comment]
```
  ==> cabal v2-install --with-compiler=/opt/homebrew/opt/ghc/bin/ghc --flags=-dynamic --jobs=4 --max-backjumps=100000 --install-method=copy --installdir=/opt/homebrew/Cellar/haskell-language-server/2.3.0.0/bin
  Error: cabal: Could not resolve dependencies:
  [__0] trying: ekg-json-0.1.0.7 (user goal)
  [__1] trying: base-4.18.0.0/installed-4.18.0.0 (dependency of ekg-json)
  [__2] trying: ghcide-2.3.0.0 (user goal)
  [__3] trying: stm-containers-1.2.0.2 (dependency of ghcide)
  [__4] next goal: stm-hamt (dependency of stm-containers)
  [__4] rejecting: stm-hamt-1.2.0.11, stm-hamt-1.2.0.10 (constraint from project
  config
  /private/tmp/haskell-language-server-20231005-27723-12xdk3f/haskell-language-server-2.3.0.0/cabal.project
  requires <1.2.0.10)
  [__4] trying: stm-hamt-1.2.0.9
  [__5] trying: primitive-0.8.0.0 (dependency of stm-hamt)
  [__6] trying: filepath-1.4.100.1/installed-1.4.100.1 (dependency of ghcide)
  [__7] next goal: exceptions (dependency of ghcide)
  [__7] rejecting: exceptions-0.10.7/installed-0.10.7 (conflict: stm-hamt =>
  transformers>=0.5 && <0.6, exceptions =>
  transformers==0.6.1.0/installed-0.6.1.0)
  [__7] rejecting: exceptions-0.10.7, exceptions-0.10.5, exceptions-0.10.4,
  exceptions-0.10.3, exceptions-0.10.2, exceptions-0.10.1, exceptions-0.10.0,
  exceptions-0.8.3, exceptions-0.8.2.1, exceptions-0.8.1, exceptions-0.8.0.2,
  exceptions-0.8.0.1, exceptions-0.8, exceptions-0.7, exceptions-0.6.1,
  exceptions-0.6, exceptions-0.5, exceptions-0.4, exceptions-0.3.3.1,
  exceptions-0.3.3, exceptions-0.3.2, exceptions-0.3.1, exceptions-0.3,
  exceptions-0.2, exceptions-0.1.1, exceptions-0.1.0.1, exceptions-0.1,
  exceptions-0.10.6, exceptions-0.9.0 (conflict: filepath =>
  exceptions==0.10.7/installed-0.10.7)
  [__7] fail (backjumping, conflict set: exceptions, filepath, ghcide, stm-hamt)
  After searching the rest of the dependency tree exhaustively, these were the
  goals I've had most trouble fulfilling: filepath, exceptions, ghcide,
  primitive, stm-containers, stm-hamt, base, ekg-json
  Try running with --minimize-conflict-set to improve the error message.
```

**Comment by iMichka** [comment]
I have a fix for that. Will push it on the next free long runner slot.

**Comment by iMichka** [comment]
I have a cleaner fix for the in replace in `haskell-language-server`, did not have time to prepare it before this run (I really wanted that long timeout build slot).

If this PR is green it can be merged, and I'll switch to this as soon as possible
https://github.com/haskell/haskell-language-server/pull/3829

**Review by fxcoudert — APPROVED** [review]

**Comment by iMichka** [comment]
Added the ci-no-fail fast label back. There was un unrelated error, rerunning ... :(

```
  ==> Testing pinboard-notes-backup
  ==> /opt/homebrew/Cellar/pinboard-notes-backup/1.0.5.6/bin/pnbackup Notes.sqlite 2>&1
  ==> /opt/homebrew/Cellar/pinboard-notes-backup/1.0.5.6/bin/pnbackup -t token Notes.sqlite 2>&1
  Error: pinboard-notes-backup: failed
  An exception occurred within a child process:
    Minitest::Assertion: Expected /HTTP\ 500\ response/ to match "Downloading the list of notes...\npnbackup: Error while communicating with the Pinboard server: InternalException (HandshakeFailed Error_EOF)\n".
```

**Review by ZhongRuoyu — APPROVED** [review]
Thanks for moving this along @iMichka!

**Review by p-linnane — APPROVED** [review]
The test for `agda` failed on Intel Sonoma. I have a version bump PR open for it right now, so we can follow up there. Merging.

Amazing work @iMichka!

**Comment by fxcoudert** [comment]
Thank you so much @iMichka for this great effort!

**Pull request merged**